File: avalon/composition/impl/src/java/org/apache/avalon/composition/model/impl/DefaultServiceRepository.java
/*

 ============================================================================
                   The Apache Software License, Version 1.1
 ============================================================================

 Copyright (C) 1999-2002 The Apache Software Foundation. All rights reserved.

 Redistribution and use in source and binary forms, with or without modifica-
 tion, are permitted provided that the following conditions are met:

 1. Redistributions of  source code must  retain the above copyright  notice,
    this list of conditions and the following disclaimer.

 2. Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution.

 3. The end-user documentation included with the redistribution, if any, must
    include  the following  acknowledgment:  "This product includes  software
    developed  by the  Apache Software Foundation  (http://www.apache.org/)."
    Alternately, this  acknowledgment may  appear in the software itself,  if
    and wherever such third-party acknowledgments normally appear.

 4. The names "Jakarta", "Apache Avalon", "Avalon Framework" and
    "Apache Software Foundation"  must not be used to endorse or promote
    products derived  from this  software without  prior written
    permission. For written permission, please contact apache@apache.org.

 5. Products  derived from this software may not  be called "Apache", nor may
    "Apache" appear  in their name,  without prior written permission  of the
    Apache Software Foundation.

 THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED WARRANTIES,
 INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 FITNESS  FOR A PARTICULAR  PURPOSE ARE  DISCLAIMED.  IN NO  EVENT SHALL  THE
 APACHE SOFTWARE  FOUNDATION  OR ITS CONTRIBUTORS  BE LIABLE FOR  ANY DIRECT,
 INDIRECT, INCIDENTAL, SPECIAL,  EXEMPLARY, OR CONSEQUENTIAL  DAMAGES (INCLU-
 DING, BUT NOT LIMITED TO, PROCUREMENT  OF SUBSTITUTE GOODS OR SERVICES; LOSS
 OF USE, DATA, OR  PROFITS; OR BUSINESS  INTERRUPTION)  HOWEVER CAUSED AND ON
 ANY  THEORY OF LIABILITY,  WHETHER  IN CONTRACT,  STRICT LIABILITY,  OR TORT
 (INCLUDING  NEGLIGENCE OR  OTHERWISE) ARISING IN  ANY WAY OUT OF THE  USE OF
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

 This software  consists of voluntary contributions made  by many individuals
 on  behalf of the Apache Software  Foundation. For more  information on the
 Apache Software Foundation, please see <http://www.apache.org/>.

*/

package org.apache.avalon.composition.model.impl;

import java.util.Iterator;
import java.util.List;

import org.apache.avalon.composition.model.ServiceUnknownException;
import org.apache.avalon.composition.model.ServiceRepository;
import org.apache.avalon.framework.Version;
import org.apache.avalon.framework.logger.Logger;
import org.apache.avalon.meta.info.ReferenceDescriptor;
import org.apache.avalon.meta.info.Service;

/**
 * A service repository provides support for the retrival
 * of service defintions.
 *
 * @author <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
 * @version $Revision: 1.2 $ $Date: 2003/10/28 20:21:00 $
 */
public class DefaultServiceRepository implements ServiceRepository
{
    //==============================================================
    // immutable state
    //==============================================================

    /**
     * The logging channel.
     */
    private final Logger m_logger;

    /**
     * The parent service manager (may be null)
     */
    private final ServiceRepository m_parent;

    /**
     * List of service entries.
     */
    private final List m_services;

    //==============================================================
    // constructor
    //==============================================================

    /**
     * Creation of a new root service manager.
     * @param logger the logging channel
     * @param services the list of available services
     * @exception NullPointerException if the services list is null
     */
    DefaultServiceRepository( 
      final Logger logger, final List services ) throws NullPointerException
    {
        this( logger, null, services );
    }

    /**
     * Creation of a new service manager.
     * @param logger the logging channel
     * @param parent the parent type manager
     * @param services the list of available services
     * @exception NullPointerException if the services list is null
     */
    DefaultServiceRepository( 
      final Logger logger, final ServiceRepository parent, final List services ) 
      throws NullPointerException
    {
        if( services == null )
        {
            throw new NullPointerException( "services" );
        }
        m_parent = parent;
        m_services = services;
        m_logger = logger;
    }

    //==============================================================
    // implemetation
    //==============================================================

    /**
     * Locate a {@link Service} instances associated with the
     * supplied classname and version. If a service defintion is not
     * found locally, the implementation redirects the request to
     * the parent service manager.
     *
     * @param classname the service class name
     * @param version the service version
     * @return the service matching the supplied classname and version.
     * @exception UnknownServiceException if a matching service cannot be found
     */
    public Service getService( final String classname, final Version version ) 
      throws ServiceUnknownException
    {
        return getService( new ReferenceDescriptor( classname, version ) );
    }

    /**
     * Locate a {@link Service} instances associated with the
     * supplied referecne descriptor. If a service defintion is not
     * found locally, the implementation redirects the request to
     * the parent service manager.
     *
     * @param reference the reference descriptor
     * @return the service matching the supplied descriptor.
     * @exception UnknownServiceException if a matching service cannot be found
     */
    public Service getService( final ReferenceDescriptor reference ) 
      throws ServiceUnknownException
    {
        Service service = getLocalService( reference );
        if( service == null )
        {
            if( m_parent != null )
            {
                return m_parent.getService( reference );
            } 
            else
            {
                final String error = "Unknown service defintion: " + reference;
                throw new ServiceUnknownException( error );
            }
        }
        return service;
    }

    private Service getLocalService( final ReferenceDescriptor reference )
    {
        Iterator iterator = m_services.iterator();
        while( iterator.hasNext() )
        {
            Service service = (Service) iterator.next();
            if( service.equals( reference ) )
            {
                return service;
            }
        }
        return null;
    }

    protected Logger getLogger()
    {
        return m_logger;
    }
}

File: avalon-sandbox/avalon-net/Composition/ModelDefault/DefaultServiceRepository.cs
/* 
* Copyright 2003-2004 The Apache Software Foundation
* Licensed  under the  Apache License,  Version 2.0  (the "License");
* you may not use  this file  except in  compliance with the License.
* You may obtain a copy of the License at 
* 
*   http://www.apache.org/licenses/LICENSE-2.0
* 
* Unless required by applicable law or agreed to in writing, software
* distributed  under the  License is distributed on an "AS IS" BASIS,
* WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
* implied.
* 
* See the License for the specific language governing permissions and
* limitations under the License.
*/

namespace Apache.Avalon.Composition.Model.Default
{
	using System;

	using Apache.Avalon.Framework;
	using Apache.Avalon.Meta;
	using Apache.Avalon.Composition.Model;
	
	/// <summary> A service repository provides support for the retrival
	/// of service defintions.
	/// 
	/// </summary>
	/// <author>  <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
	/// </author>
	/// <version>  $Revision: 1.1 $ $Date: 2004/02/28 22:15:42 $
	/// </version>
	public class DefaultServiceRepository : IServiceRepository
	{

		//==============================================================
		// immutable state
		//==============================================================
		
		/// <summary> The logging channel.</summary>
		//UPGRADE_NOTE: Final was removed from the declaration of 'm_logger '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		private ILogger m_logger;
		
		/// <summary> The parent service manager (may be null)</summary>
		//UPGRADE_NOTE: Final was removed from the declaration of 'm_parent '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		private IServiceRepository m_parent;
		
		/// <summary> List of service entries.</summary>
		//UPGRADE_NOTE: Final was removed from the declaration of 'm_services '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		//UPGRADE_ISSUE: Class hierarchy differences between ''java.util.List'' and ''System.Collections.IList'' may cause compilation errors. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1186"'
		private System.Collections.IList m_services;
		
		//==============================================================
		// constructor
		//==============================================================
		
		/// <summary> Creation of a new root service manager.</summary>
		/// <param name="logger">the logging channel
		/// </param>
		/// <param name="services">the list of available services
		/// </param>
		/// <exception cref=""> NullPointerException if the services list is null
		/// </exception>
		//UPGRADE_ISSUE: Class hierarchy differences between ''java.util.List'' and ''System.Collections.IList'' may cause compilation errors. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1186"'
		internal DefaultServiceRepository(ILogger logger, System.Collections.IList services):this(logger, null, services)
		{
		}
		
		/// <summary> Creation of a new service manager.</summary>
		/// <param name="logger">the logging channel
		/// </param>
		/// <param name="parent">the parent type manager
		/// </param>
		/// <param name="services">the list of available services
		/// </param>
		/// <exception cref=""> NullPointerException if the services list is null
		/// </exception>
		//UPGRADE_ISSUE: Class hierarchy differences between ''java.util.List'' and ''System.Collections.IList'' may cause compilation errors. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1186"'
		internal DefaultServiceRepository(ILogger logger, IServiceRepository parent, System.Collections.IList services)
		{
			if (services == null)
			{
				throw new System.ArgumentNullException("services");
			}
			m_parent = parent;
			m_services = services;
			m_logger = logger;
		}
		
		//==============================================================
		// implemetation
		//==============================================================
		
		/// <summary> Locate a {@link Service} instances associated with the
		/// supplied typename and version. If a service defintion is not
		/// found locally, the implementation redirects the request to
		/// the parent service manager.
		/// 
		/// </summary>
		/// <param name="typename">the service class name
		/// </param>
		/// <param name="version">the service version
		/// </param>
		/// <returns> the service matching the supplied typename and version.
		/// </returns>
		/// <exception cref=""> UnknownServiceException if a matching service cannot be found
		/// </exception>
		public virtual Service GetService(Type type)
		{
			return GetService(new ReferenceDescriptor(type));
		}
		
		/// <summary> Locate a {@link Service} instances associated with the
		/// supplied referecne descriptor. If a service defintion is not
		/// found locally, the implementation redirects the request to
		/// the parent service manager.
		/// 
		/// </summary>
		/// <param name="reference">the reference descriptor
		/// </param>
		/// <returns> the service matching the supplied descriptor.
		/// </returns>
		/// <exception cref=""> UnknownServiceException if a matching service cannot be found
		/// </exception>
		public virtual Service GetService(ReferenceDescriptor reference)
		{
			Service service = GetLocalService(reference);
			if (service == null)
			{
				if (m_parent != null)
				{
					return m_parent.GetService(reference);
				}
				else
				{
					//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
					String error = "Unknown service defintion: " + reference;
					throw new ServiceUnknownException(error);
				}
			}
			return service;
		}
		
		protected virtual internal ILogger Logger
		{
			get
			{
				return m_logger;
			}
			
		}

		private Service GetLocalService(ReferenceDescriptor reference)
		{
			System.Collections.IEnumerator iterator = m_services.GetEnumerator();
			//UPGRADE_TODO: Method 'java.util.Iterator.hasNext' was converted to 'System.Collections.IEnumerator.MoveNext' which has a different behavior. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1073_javautilIteratorhasNext"'
			while (iterator.MoveNext())
			{
				//UPGRADE_TODO: Method 'java.util.Iterator.next' was converted to 'System.Collections.IEnumerator.Current' which has a different behavior. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1073_javautilIteratornext"'
				/* Service service = (Service) iterator.Current;
				if (service.Equals(reference))
				{
					return service;
				}*/

				Service service = (Service) iterator.Current;
				if (service.Reference.Equals( reference ))
				{
					return service;
				}
			}
			return null;
		}
	}
}